### PR TITLE
KAFKA-16068: Use TestPlugins in ConnectorValidationIntegrationTest to silence plugin scanning errors

### PR DESCRIPTION
**Pull request by Joker-5**
[Jira](https://issues.apache.org/jira/browse/KAFKA-16068)

The ConnectorValidationIntegrationTest creates test plugins, some with erroneous behavior. In particular:
```
[2023-12-29 10:28:06,548] ERROR Failed to discover Converter in classpath: Unable to instantiate TestConverterWithPrivateConstructor: Plugin class default constructor must be public (org.apache.kafka.connect.runtime.isolation.ReflectionScanner:138) [2023-12-29 10:28:06,550]
ERROR Failed to discover Converter in classpath: Unable to instantiate TestConverterWithConstructorThatThrowsException: Failed to invoke plugin constructor (org.apache.kafka.connect.runtime.isolation.ReflectionScanner:138)
java.lang.reflect.InvocationTargetException
```
I found 4 methods with erroneous behavior:
- testConnectorHasConverterWithNoSuitableConstructor()
- testConnectorHasConverterThatThrowsExceptionOnInstantiation()
- testConnectorHasHeaderConverterWithNoSuitableConstructor()
- testConnectorHasHeaderConverterThatThrowsExceptionOnInstantiation()

These plugins should be eliminated from the classpath, so that the errors do not appear in unrelated tests. Instead, plugins with erroneous behavior should only be present in the TestPlugins, so that tests can opt-in to loading them.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java [hunk 1]
```diff
@@ -331,7 +334,7 @@ public void testConnectorHasAbstractConverter() throws InterruptedException {
     @Test
     public void testConnectorHasConverterWithNoSuitableConstructor() throws InterruptedException {
         Map<String, String> config = defaultSinkConnectorProps();
-        config.put(KEY_CONVERTER_CLASS_CONFIG, TestConverterWithPrivateConstructor.class.getName());
+        config.put(KEY_CONVERTER_CLASS_CONFIG, TestPlugins.TestPlugin.ALWAYS_THROW_EXCEPTION.className());
```
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: I think we should use `BAD_PACKAGING_DEFAULT_CONSTRUCTOR_PRIVATE_CONNECTOR` in this case as well?

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java [hunk 2]
```diff
@@ -343,7 +346,7 @@ public void testConnectorHasConverterWithNoSuitableConstructor() throws Interrup
     @Test
     public void testConnectorHasConverterThatThrowsExceptionOnInstantiation() throws InterruptedException {
         Map<String, String> config = defaultSinkConnectorProps();
-        config.put(KEY_CONVERTER_CLASS_CONFIG, TestConverterWithConstructorThatThrowsException.class.getName());
+        config.put(KEY_CONVERTER_CLASS_CONFIG, TestPlugins.TestPlugin.ALWAYS_THROW_EXCEPTION.className());
```
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: We could probably use `BAD_PACKAGING_DEFAULT_CONSTRUCTOR_THROWS_CONNECTOR` and in `testConnectorHasHeaderConverterThatThrowsExceptionOnInstantiation` as well?

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java [hunk 3]
```diff
@@ -69,6 +69,9 @@ public static void setup() {
         Map<String, String> workerProps = new HashMap<>();
         workerProps.put(GROUP_ID_CONFIG, WORKER_GROUP_ID);
 
+        // Work around a circular-dependency in TestPlugins.
+        TestPlugins.pluginPath();
```
Comment (vamossagar12) [inline connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: Why do we need this?

**Review by vamossagar12 — COMMENTED** [review]
Thanks for the PR @Joker-5 . I left a couple of comments, let me know if those make sense.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java (hunk 1 above)
Comment (Joker-5) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: oops, I misread the code, now I've corrected it.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java (hunk 2 above)
Comment (Joker-5) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: At the beginning I used `BAD_PACKAGING_DEFAULT_CONSTRUCTOR_THROWS_CONNECTOR` in the case, but I thought `ALWAYS_THROW_EXCEPTION` has stronger sematics, that's why I changed my mind.

But now I think about the case again, I think `BAD_PACKAGING_DEFAULT_CONSTRUCTOR_THROWS_CONNECTOR ` has fine-grain sematics which is more suitable for the case.

I've changed it. Thanks so much for pointing out that.

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java (hunk 3 above)
Comment (Joker-5) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: @vamossagar12 If we don't add this, methods which use `TestPlugins`  will have erroneous behavior. In particular:
```
[2024-04-02 19:16:25,977] ERROR Could not set up plugin test jars (org.apache.kafka.connect.runtime.isolation.TestPlugins:258)
java.lang.NullPointerException
	at org.apache.kafka.connect.runtime.isolation.TestPlugins$TestPlugin.values(TestPlugins.java:69)
	at org.apache.kafka.connect.runtime.isolation.TestPlugins.<clinit>(TestPlugins.java:251)
	at org.apache.kafka.connect.runtime.isolation.TestPlugins$TestPlugin.<clinit>(TestPlugins.java:128)
	at org.apache.kafka.connect.integration.ConnectorValidationIntegrationTest.testConnectorHasConverterWithNoSuitableConstructor(ConnectorValidationIntegrationTest.java:337)
...
```
I found the reason behind it:

>TL;DR
The whole error occurred because the circle: inner class(some field depend on outer class in the \<clinit> method) -> outer class(some line in static code block depend on inner class in the \<clinit> method) -> inner class

1. There's a `private final Predicate<String> removeRuntimeClasses` field in the inner class `org.apache.kafka.connect.runtime.isolation.TestPlugins.TestPlugin`.
2. Some enums such as `BAD_PACKAGING_MISSING_SUPERCLASS` in (1)'s field will depend on `private static final Predicate<String> REMOVE_CLASS_FILTER = s -> s.contains("NonExistentInterface")` in the outer class(`org.apache.kafka.connect.runtime.isolation.TestPlugins`).
3. When test mothod such as `testConnectorHasConverterWithNoSuitableConstructor` is running, JVM will call \<clinit> method to initialize the inner class(`org.apache.kafka.connect.runtime.isolation.TestPlugins.TestPlugin`).
4. But some enum use the field(`removeRuntimeClasses`) in the outer class(`private static final Predicate<String> REMOVE_CLASS_FILTER = s -> s.contains("NonExistentInterface")`), so at that time JVM has to call \<clinit> method to initialize the outer class(`org.apache.kafka.connect.runtime.isolation.TestPlugins`).
5. In the outer class, there's a static code block which used the inner class. But the inner class have not been initialized, the circular-dependency just happened, which result in that error log.
```java
    static {
        Throwable err = null;
        Map<String, Path> pluginJars = new HashMap<>();
        try {
            for (TestPlugin testPlugin : TestPlugin.values()) { // see this line
                if (pluginJars.containsKey(testPlugin.resourceDir())) {
                    log.debug("Skipping recompilation of " + testPlugin.resourceDir());
                }
                pluginJars.put(testPlugin.resourceDir(), createPluginJar(testPlugin.resourceDir(), testPlugin.removeRuntimeClasses()));
            }
        } catch (Throwable e) {
            log.error("Could not set up plugin test jars", e);
            err = e;
        }
        PLUGIN_JARS = Collections.unmodifiableMap(pluginJars);
        INITIALIZATION_EXCEPTION = err;
    }
```

So that's why I added the code.

Comment (vamossagar12) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: Thanks for the explanation. hmm I ran the tests in `ConnectorValidationIntergationTest` without this one line but including all others and the tests are passing for me locally. I created a branch (draft, we can close it once we have validated the behaviour) https://github.com/apache/kafka/pull/15653 with the changes. We can monitor if the circular dependency related errors show up there or not.

Comment (Joker-5) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: > Thanks for the explanation. hmm I ran the tests in `ConnectorValidationIntergationTest` without this one line but including all others and the tests are passing for me locally. I created a branch (draft, we can close it once we have validated the behaviour) #15653 with the changes. We can monitor if the circular dependency related errors show up there or not.

@vamossagar12, I checked the result from #15653 which did not have related errors, so I removed this one line.
Maybe there's some problem in my local env, I will try to investigate the root cause :)

**Review by vamossagar12 — APPROVED** [review]
Thanks for the changes, LGTM

File: connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java (hunk 3 above)
Comment (gharris1727) [reply connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorValidationIntegrationTest.java]: The circular dependency is still a problem that I haven't resolved, sorry. It's fixed in this un-merged PR: #13333

**Review by gharris1727 — COMMENTED** [review]
Hi @Joker-5 Thanks for picking this up!

These tests have different semantics now. Before, the faulty test classes generated errors in the InstantiableClassValidator and ConcreteSubClassValidator. In the PR currently, the test classes can't be found, because the TestPlugin class name is not on the classpath. You can see the difference in the logs for the test:

Before:
```
Invalid value org.apache.kafka.connect.integration.ConnectorValidationIntegrationTest$TestConverterWithConstructorThatThrowsException for configuration header.converter:
Could not instantiate class
```
After:
```
Invalid value test.plugins.DefaultConstructorThrowsConnector for configuration key.converter:
Class test.plugins.DefaultConstructorThrowsConnector could not be found
```

Even if we change the setup to include this line to add the jars containing test.plugins.* to the plugin.path:
```
workerProps.put(WorkerConfig.PLUGIN_PATH_CONFIG, TestPlugins.pluginPathJoined())
```
the error messages are still different, because:
1. The plugin can't be found
    1. It's in a different classloader (MonitorableSinkConnector is on the classpath, the faulty plugins are now on the plugin.path)
    2. Plugins with faulty constructors are hidden from plugin discovery #13467
    3. Plugins which aren't discoverable cannot be found via DelegatingClassLoader
2. If it could be found, it's of the wrong type
    1. The configuration key is `*_CONVERTER_*`
    2. The new plugins are `*_CONNECTOR`

I think this should still be an integration test of plugin isolation + connector validation + config def + InstantiableClassValidator, so we should keep iterating on this PR rather than rewriting these as unit tests.

There is a NoDefaultConstructorConverter that you could re-use, and you'd need to add a valid connector to the bad-packaging jar to use in-place of MonitorableSourceConnector. That would also let you verify in the other tests that the plugin.path is set up correctly.
We could also not use the bad-packaging jar, and give this test it's own jar. That's up to you.

**Comment by gharris1727** [comment]
This has been superseded by #16647 and so is no longer necessary.
